### PR TITLE
Add tests to virtualbox driver

### PR DESCRIPTION
**Pull request by dgageot**
This is an ongoing PR aiming at increasing test coverage for the virtualbox driver

### CODE REVIEW

**Comment by nathanleclaire** [comment]
Similar to in the Google Cloud tags PR, the testing approach here needs to be different than simply making the `vbm` etc. functions into variables and setting them for the tests. That approach won't scale as we expand out test coverage and want to be able to cover more complex functions such as `Create` (imagine trying to set the results of that function correctly for every single time it gets called in all the different possible paths of a `Create`).

Instead, the first steps towards testing this are to expose the `VBoxManage` operations through an `interface` which we then make a member of the `Driver` struct.  Something like:

```go
type VBoxManager interface {
    vbmOutErr (string, string, error)
}

type VBoxCmdManager struct {
    // Params that make sense for the concrete struct to interact
    // with 'VBoxManage'
}

// vbmOut and vbm could be methods defined on VBoxCmdManager,
// since they loosely wrap vbmOutErr they largely won't need to be over-ridden in 
// tests.

// implementation...
```

And `Driver` embeds / uses it, like so:

```go
type Driver struct {
    *VBoxCmdManager
}

// make sure to initialize VBoxCmdManager in NewDriver function
```

(consequently all calls to `vbm` etc. in the driver file become `d.Vbm` etc...)

Consequently, we could mock out calls to `VBoxManage` in the tests doing something like:

```go
type VBoxCmdResult struct {
	stdout, stderr string
	err            error
}

type VBoxManageMock struct {
	expectedOutErr map[string]*VBoxCmdResult
}

func (vbmm VBoxManageMock) vbmOutErr(args ...string) (string, string, error) {
	key := strings.Join(args, " ")
	result := vbmm.expectedOutErr[key]
	return result.stdout, result.stderr, result.err
}

func TestVbmVersion(t *testing.T) {
	cmdResultMap := map[string]*VBoxCmdResult{
		"--version": &VBoxCmdResult{
			stdout: "5.0.6r103037",
			stderr: "",
			err:    nil,
		},
	}

	vboxManageMock := &VBoxManageMock{
		expectedOutErr: cmdResultMap,
	}

	driver := NewDriver("testvm", "/store/path")
	driver.VBoxCmdManager = vboxManageMock

	// etc...
}
```

(could be revised to be more fancy and/or expanded in scope as needed)

I know that it's somewhat annoying to re-architect at the same time as tests are being written, but it is really important to ensure a sustainable testing workflow in the future.  If you have any questions, ask away.  I'd rather take the time to do it right.

**Comment by dgageot** [comment]
@nathanleclaire Please take a look at the tests that now use an interface to mock the communication with virtualbox. wdyt?

**Comment by nathanleclaire** [comment]
Looking better - I have some nitpicks and suggestions I'll write inline.

File: drivers/virtualbox/virtualbox.go [hunk 1]
```diff
@@ -62,8 +63,10 @@ func init() {
 	})
 }
 
+// NewDriver creates a new VirtualBox driver with default settings.
 func NewDriver(hostName, storePath string) *Driver {
 	return &Driver{
+		VBoxManager: &VBoxCmdManager{},
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox.go]: :+1:

File: drivers/virtualbox/virtualbox.go [hunk 2]
```diff
@@ -182,11 +185,7 @@ func (d *Driver) SetConfigFromFlags(flags drivers.DriverOptions) error {
 
 func (d *Driver) PreCreateCheck() error {
 	// Check that VBoxManage exists and works
-	if err := vbm(); err != nil {
-		return err
-	}
-
-	return nil
+	return vbm()
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox.go]: :+1:

File: drivers/virtualbox/vbm.go [hunk 3]
```diff
@@ -56,6 +56,19 @@ func setVBoxManageCmd() string {
 	return cmd
 }
 
+// VBoxManager defines the interface to communicate to VirtualBox.
+type VBoxManager interface {
+	queryVirtualBox(args ...string) (string, string, error)
```
Comment (nathanleclaire) [inline drivers/virtualbox/vbm.go]: Nitpick:  I'd call this `vbmOutErr` for consistency with the package-level function and minimal search-and-replace effort.

File: drivers/virtualbox/vbm.go [hunk 4]
```diff
@@ -56,6 +56,19 @@ func setVBoxManageCmd() string {
...
+}
+
+// VBoxCmdManager will communicate with VirtualBox through the commandline
+// using `VBoxManage`.
+type VBoxCmdManager struct{}
+
+func (v *VBoxCmdManager) queryVirtualBox(args ...string) (string, string, error) {
+	return vbmOutErr(args...)
```
Comment (nathanleclaire) [inline drivers/virtualbox/vbm.go]: Looks good.  Longer-termish, we should think about embedding all related logic (e.g. finding the path of `VBoxManage`) into `VBoxCmdManager` directly, but this is an acceptable solution to get the ball rolling on testing.

File: drivers/virtualbox/virtualbox.go [hunk 5]
```diff
@@ -499,7 +498,7 @@ func (d *Driver) Kill() error {
 }
 
 func (d *Driver) GetState() (state.State, error) {
-	stdout, stderr, err := vbmOutErr("showvminfo", d.MachineName,
+	stdout, stderr, err := d.queryVirtualBox("showvminfo", d.MachineName,
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox.go]: As mentioned above, I know it's nitpicky, but my preference would be to have this as `d.vbmOutErr(`for now.

File: drivers/virtualbox/virtualbox_test.go [hunk 6]
```diff
@@ -29,3 +226,27 @@ func TestGetRandomIPinSubnet(t *testing.T) {
...
 	}
 }
+
+func newDriver(name string) *Driver {
+	return NewDriver(name, "")
+}
+
+// create parses the provided args and tries to configure the driver for
+// a `create` action.
+func create(args []string) (*Driver, error) {
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: I'm still adamant about removing this in favor of just creating the `Driver` structs directly.  All references to `codegangsta/cli` in the driver code are being removed in the plugins work so we shouldn't introduce any more, and we should stick to testing strictly what's in this package.

File: drivers/virtualbox/virtualbox_test.go [hunk 7]
```diff
@@ -29,3 +226,27 @@ func TestGetRandomIPinSubnet(t *testing.T) {
 		t.Fatalf("expected third octet of %d; received %d", testIP[2], newIP[2])
 	}
 }
+
+func newDriver(name string) *Driver {
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: I'd prefer to call this something like `newTestDriver` for clarity.

File: drivers/virtualbox/virtualbox_test.go [hunk 8]
```diff
@@ -1,10 +1,207 @@
...
+}
+
+func (v *VBoxManagerMock) queryVirtualBox(args ...string) (string, string, error) {
+	if strings.Join(args, " ") == v.expectedArgs {
+		return v.output, v.stdErr, v.err
+	}
+	return "", "", errors.New("Invalid args")
+}
+
+func TestStateRunning(t *testing.T) {
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: Why not consolidate these `GetState` tests into one if you using a [test table](http://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go)?  Something like:

```go
package main

import "testing"

func TestGetState(t *testing.T) {
	cases := []struct {
		driver        Driver
		vbManager     VBoxManager
		expectedState state.State
		expectedErr   error
	}{
		{
			driver: newDriver("default"),
			vbManager: &VBoxManagerMock{
				expectedArgs: "showvminfo default --machinereadable",
				output:       `VMState="aborted"`,
			},
			expectedState: state.Stopped,
			expectedErr:   nil,
		},
		{
			driver: newDriver("default"),
			vbManager: &VBoxManagerMock{
				expectedArgs: "showvminfo default --machinereadable",
				output:       `VMState="running"`,
			},
			expectedState: state.Running,
			expectedErr:   nil,
		},
	}

	for _, c := range cases {
		driver := c.driver
		driver.VBoxManager = c.vbManager
		machineState, err := driver.GetState()
		assert.Equal(t, state.Stopped, machineState)
		assert.Equal(t, err)
	}

}
```

Does it make sense?

File: drivers/virtualbox/virtualbox_test.go [hunk 9]
```diff
@@ -1,10 +1,207 @@
...
+	assert.Equal(t, driver.HostOnlyCIDR, "192.168.99.1/8")
+	assert.Equal(t, driver.HostOnlyNicType, "test-nic-type")
+	assert.Equal(t, driver.HostOnlyPromiscMode, "accept")
+	assert.Equal(t, driver.NoShare, true)
+}
+
+func TestSSHHostname(t *testing.T) {
+	hostname, err := newDriver("default").GetSSHHostname()
+
+	assert.Equal(t, "127.0.0.1", hostname)
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: I still don't feel particularly strongly that we need tests to check for these values, but if you insist, please move the magic strings (e.g. `127.0.0.1`) to package-level `const` so that we only need to update in one place if we change them.

File: drivers/virtualbox/virtualbox_test.go [hunk 10]
```diff
@@ -1,10 +1,207 @@
...
+	driver := newDriver("default")
+	driver.VBoxManager = &VBoxManagerMock{
+		expectedArgs: "showvminfo default --machinereadable",
+		output:       `VMState="whatever"`,
+	}
+
+	machineState, err := driver.GetState()
+
+	assert.Equal(t, state.None, machineState)
+	assert.NoError(t, err) // Really?
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: Heh, "Really" as in "Seriously, we don't return an error in this case?"  I agree it should raise an error.

File: drivers/virtualbox/virtualbox_test.go [hunk 11]
```diff
@@ -1,10 +1,207 @@
...
+}
+
+func TestStateOfMachineNotFound(t *testing.T) {
+	driver := newDriver("default")
+	driver.VBoxManager = &VBoxManagerMock{
+		expectedArgs: "showvminfo default --machinereadable",
+		stdErr:       "Could not find a registered machine named 'unknown'",
+		err:          errors.New("Bug"),
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: Maybe a good candidate for a typed error

File: drivers/virtualbox/virtualbox_test.go [hunk 12]
```diff
@@ -1,10 +1,207 @@
...
+}
+
+type VBoxManagerMock struct {
+	expectedArgs string
+	output       string
+	stdErr       string
+	err          error
+}
+
+func (v *VBoxManagerMock) queryVirtualBox(args ...string) (string, string, error) {
```
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: This is a decent start but it needs to be a little more robust.  I would prefer that the `VBoxManagerMock` had an internal `map[string]*CmdResult` and the code for this method would be more like:

```go
	cmdResult, ok := v.CmdResults[strings.Join(args, " ")]
	if !ok {
		return "", "", errors.New("Command not found in test map")
	}
	return cmdResult.stdout, cmdResult.stderr, cmdResult.err
```

**Comment by nathanleclaire** [comment]
Wrote quite a few nitpicks and asks, but looking a lot better.

File: drivers/virtualbox/vbm.go (hunk 3 above)
Comment (dgageot) [inline drivers/virtualbox/vbm.go]: good

File: drivers/virtualbox/virtualbox_test.go (hunk 6 above)
Comment (dgageot) [inline drivers/virtualbox/virtualbox_test.go]: Then what's your idea on testing that flag definition and flag reading are aligned and working as expected. This can't be done all in integration tests

File: drivers/virtualbox/virtualbox_test.go (hunk 7 above)
Comment (dgageot) [inline drivers/virtualbox/virtualbox_test.go]: ok

File: drivers/virtualbox/virtualbox_test.go (hunk 8 above)
Comment (dgageot) [inline drivers/virtualbox/virtualbox_test.go]: It makes sense. The table would still be huge and would contain a lot of duplication. I'd rather improve the individual tests first and then we'll see about the test table.

File: drivers/virtualbox/virtualbox_test.go (hunk 9 above)
Comment (dgageot) [inline drivers/virtualbox/virtualbox_test.go]: I prefer not. Unit tests are here to protect us from a change in the code. Sharing the const between the test and the code make it useless.
Somebody makes a mistake and changes the `127.0.0.1` in the code -> the test will catch it.
Somebody extract the `127.0.0.1` into a method that computes the value -> the test will catch it.

File: drivers/virtualbox/virtualbox_test.go (hunk 10 above)
Comment (dgageot) [inline drivers/virtualbox/virtualbox_test.go]: yes ;-) I'll cover the code as is for now until I get more confidence in changing the actual code

File: drivers/virtualbox/virtualbox_test.go (hunk 11 above)
Comment (dgageot) [inline drivers/virtualbox/virtualbox_test.go]: As I said, I'm always reluctant in sharing consts between tests and code. It make the test less useful. Or maybe I didn't understand what you said.

File: drivers/virtualbox/virtualbox_test.go (hunk 12 above)
Comment (dgageot) [inline drivers/virtualbox/virtualbox_test.go]: We'll see when we get there!

File: drivers/virtualbox/vbm.go [hunk 13]
```diff
@@ -15,8 +15,6 @@ import (
 )
 
 var (
-	reVMNameUUID      = regexp.MustCompile(`"(.+)" {([0-9a-f-]+)}`)
```
Comment (nathanleclaire) [inline drivers/virtualbox/vbm.go]: Why are these removed?

Comment (dgageot) [inline drivers/virtualbox/vbm.go]: Because it's dead code! It's used nowhere.

Comment (nathanleclaire) [inline drivers/virtualbox/vbm.go]: Fair, but we should run by github.com/boot2docker/boot2docker-cli to see if it was something relevant which was accidentally left out in the port to Machine.

File: drivers/virtualbox/virtualbox_test.go (hunk 8 above)
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: Go ahead and make it into a test table if you don't mind, please.

File: drivers/virtualbox/virtualbox_test.go (hunk 6 above)
Comment (nathanleclaire) [inline drivers/virtualbox/virtualbox_test.go]: > Then what's your idea on testing that flag definition and flag reading are aligned and working as expected. This can't be done all in integration tests

We can test `SetConfigFromFlags` directly.  Just have to pass in an instance of `DriverOpts` which we mock out to be whatever we want.

File: drivers/virtualbox/vbm.go (hunk 13 above)
Comment (nathanleclaire) [inline drivers/virtualbox/vbm.go]: https://github.com/boot2docker/boot2docker-cli/blob/9a2606673efcfa282fb64a5a5c9e1b2f89d86fb4/virtualbox/machine.go#L425

**Comment by dgageot** [comment]
Hi @nathanleclaire I took all your feedback into account. Can you take a look?

File: drivers/virtualbox/vbm.go (hunk 13 above)
Comment (nathanleclaire) [inline drivers/virtualbox/vbm.go]: At any rate, maybe it's right to just take it out and add it back in later if we see fit.

File: drivers/virtualbox/vbm.go [hunk 14]
```diff
@@ -11,62 +11,51 @@ import (
...
-func vbmOut(args ...string) (string, error) {
-	stdout, _, err := vbmOutErr(args...)
+func (v *VBoxCmdManager) vbmOut(args ...string) (string, error) {
+	stdout, _, err := v.vbmOutErr(args...)
 	return stdout, err
 }
 
-func vbmOutErr(args ...string) (string, string, error) {
+func (v *VBoxCmdManager) vbmOutErr(args ...string) (string, string, error) {
+	once.Do(func() {
```
Comment (nathanleclaire) [inline drivers/virtualbox/vbm.go]: Instead of `once.Do`, which is meant to be used as a concurrency primitive, I'd suggest just putting this in an `init` function in the module:

```go
func init () {
    vboxManageCmd = detectVBoxManageCmd()
}
```

Comment (dgageot) [inline drivers/virtualbox/vbm.go]: I'm going to put it back where it was before, ie:

```
var (
	vboxManageCmd = detectVBoxManageCmd()
)
```

Once this PR is merged, I'll try to move it to `VBoxCmdManager struct` because that's really where it belongs.

**Comment by dgageot** [comment]
@nathanleclaire feedback applied. There's more to be done on that subject but I think it's a good start.

**Comment by dmp42** [comment]
Ok, thanks @nathanleclaire for the review and thanks @dgageot for this! Keep them coming!

**Comment by nathanleclaire** [comment]
:+1: :+1: :+1:
